File: reference-implementations/go/extract.go
package main

import (
	"fmt"
	"io"
	"os"
	"regexp"
	"sort"
	"strings"
	"time"

	toml "github.com/pelletier/go-toml/v2"
)

var bareTomlKeyPattern = regexp.MustCompile(`^[A-Za-z0-9_-]+$`)

func extract(documentPath, schemaPath string, out, errOut io.Writer) int {
	document, err := parseTOMLFile(documentPath)
	if err != nil {
		fmt.Fprintln(errOut, err)
		return 1
	}
	if err := os.WriteFile(schemaPath, []byte(generateSchema(document)), 0o644); err != nil {
		fmt.Fprintln(errOut, err)
		return 2
	}
	fmt.Fprintf(out, "Extracted schema to %s\n", schemaPath)
	return 0
}

func generateSchema(document map[string]any) string {
	var schema strings.Builder
	schema.WriteString("[toml-schema]\n")
	fmt.Fprintf(&schema, "version = %q\n\n", currentTomlSchemaVersion)
	schema.WriteString("[elements]\n")
	for _, key := range sortedKeys(document) {
		if key == "toml-schema" {
			continue
		}
		appendDefinition(&schema, []string{"elements", key}, document[key])
	}
	return schema.String()
}

func appendDefinition(schema *strings.Builder, path []string, value any) {
	schema.WriteString("\n[")
	for i, part := range path {
		if i > 0 {
			schema.WriteByte('.')
		}
		schema.WriteString(encodeTomlKey(part))
	}
	schema.WriteString("]\n")
	typeName := schemaType(value)
	fmt.Fprintf(schema, "type = %q\n", typeName)
	if typeName == string(TypeArray) {
		fmt.Fprintf(schema, "arraytype = %q\n", inferArrayType(value.([]any)))
	}
	if table, ok := value.(map[string]any); ok {
		for _, childKey := range sortedKeys(table) {
			childPath := append(append([]string{}, path...), childKey)
			appendDefinition(schema, childPath, table[childKey])
		}
	}
}

func schemaType(value any) string {
	switch value.(type) {
	case string:
		return string(TypeString)
	case int64:
		return string(TypeInteger)
	case float64:
		return string(TypeFloat)
	case bool:
		return string(TypeBoolean)
	case time.Time:
		return string(TypeOffsetDateTime)
	case toml.LocalDateTime:
		return string(TypeLocalDateTime)
	case toml.LocalDate:
		return string(TypeLocalDate)
	case toml.LocalTime:
		return string(TypeLocalTime)
	case []any:
		return string(TypeArray)
	case map[string]any:
		return string(TypeTable)
	default:
		return string(TypeAny)
	}
}

func inferArrayType(array []any) string {
	if len(array) == 0 {
		return string(TypeAny)
	}
	firstType := schemaType(array[0])
	for _, item := range array[1:] {
		if schemaType(item) != firstType {
			return string(TypeAny)
		}
	}
	return firstType
}

func encodeTomlKey(key string) string {
	if bareTomlKeyPattern.MatchString(key) {
		return key
	}
	var encoded strings.Builder
	encoded.WriteByte('"')
	for _, current := range key {
		switch current {
		case '\\':
			encoded.WriteString(`\\`)
		case '"':
			encoded.WriteString(`\"`)
		case '\b':
			encoded.WriteString(`\b`)
		case '\t':
			encoded.WriteString(`\t`)
		case '\n':
			encoded.WriteString(`\n`)
		case '\f':
			encoded.WriteString(`\f`)
		case '\r':
			encoded.WriteString(`\r`)
		default:
			if current < 0x20 {
				fmt.Fprintf(&encoded, `\u%04X`, current)
			} else {
				encoded.WriteRune(current)
			}
		}
	}
	encoded.WriteByte('"')
	return encoded.String()
}

func sortedKeys(table map[string]any) []string {
	keys := make([]string, 0, len(table))
	for key := range table {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	return keys
}

File: reference-implementations/rust/src/extract.rs
//! Schema extraction: turns a sample TOML document into a starter `.tosd`
//! schema. Mirrors the Java and Go implementations.

use std::fmt::Write as _;

use toml::value::Datetime;
use toml::{Table, Value};

use crate::schema::{encode_path_key, SchemaType, CURRENT_TOML_SCHEMA_VERSION};

/// Renders a TOML Schema document as a TOML string from the parsed sample document. Keys
/// are emitted in the natural order of the parsed [`Table`] (which is sorted
/// because the `toml` crate uses a `BTreeMap` by default).
pub fn generate_schema(document: &Table) -> String {
    let mut schema = String::new();
    schema.push_str("[toml-schema]\n");
    writeln!(schema, "version = \"{CURRENT_TOML_SCHEMA_VERSION}\"\n").expect("write to String");
    schema.push_str("[elements]\n");
    for (key, value) in document.iter() {
        if key == "toml-schema" {
            continue;
        }
        append_definition(&mut schema, &["elements", key], value);
    }
    schema
}

fn append_definition(schema: &mut String, path: &[&str], value: &Value) {
    schema.push_str("\n[");
    for (index, part) in path.iter().enumerate() {
        if index > 0 {
            schema.push('.');
        }
        schema.push_str(&encode_toml_key(part));
    }
    schema.push_str("]\n");
    let type_name = schema_type_of(value);
    writeln!(schema, "type = \"{}\"", type_name.schema_name()).expect("write to String");
    if type_name == SchemaType::Array {
        if let Value::Array(array) = value {
            writeln!(
                schema,
                "arraytype = \"{}\"",
                infer_array_type(array).schema_name()
            )
            .expect("write to String");
        }
    }
    if let Value::Table(table) = value {
        for (child_key, child_value) in table.iter() {
            let mut child_path: Vec<&str> = path.iter().copied().collect();
            child_path.push(child_key);
            append_definition(schema, &child_path, child_value);
        }
    }
}

fn schema_type_of(value: &Value) -> SchemaType {
    match value {
        Value::String(_) => SchemaType::String,
        Value::Integer(_) => SchemaType::Integer,
        Value::Float(_) => SchemaType::Float,
        Value::Boolean(_) => SchemaType::Boolean,
        Value::Datetime(datetime) => schema_type_of_datetime(datetime),
        Value::Array(_) => SchemaType::Array,
        Value::Table(_) => SchemaType::Table,
    }
}

fn schema_type_of_datetime(datetime: &Datetime) -> SchemaType {
    match (
        datetime.date.is_some(),
        datetime.time.is_some(),
        datetime.offset.is_some(),
    ) {
        (true, true, true) => SchemaType::OffsetDateTime,
        (true, true, false) => SchemaType::LocalDateTime,
        (true, false, false) => SchemaType::LocalDate,
        (false, true, false) => SchemaType::LocalTime,
        _ => SchemaType::Any,
    }
}

fn infer_array_type(array: &[Value]) -> SchemaType {
    let Some(first) = array.first() else {
        return SchemaType::Any;
    };
    let first_type = schema_type_of(first);
    for item in &array[1..] {
        if schema_type_of(item) != first_type {
            return SchemaType::Any;
        }
    }
    first_type
}

fn encode_toml_key(key: &str) -> String {
    if !key.is_empty()
        && key
            .chars()
            .all(|character| character.is_ascii_alphanumeric() || character == '_' || character == '-')
    {
        return key.to_string();
    }
    // Reuse the path-key encoder because it produces TOML-compatible quoted keys.
    encode_path_key(key)
}
